blog/12-18-22-celo-dao/12-18-22-how-to-create-dao-on-celo.md

```diff
@@ -0,0 +1,409 @@
+
+```bash
+npm install --save-dev "hardhat@^2.12.2" "@nomicfoundation/hardhat-toolbox@^2.0.0" "solidity-coverage"
+```
+
+We will also need to install OpenZeppelin Library as dependency for our project:
+
+```bash
+npm install "@openzeppelin/contracts"
+```
+
+Let’s see what these can help us with:
+- `hardhat`: Hardhat is an extensible developer tool that helps smart contract developers increase productivity by reliably bringing together the tools they want.
+- `@nomicfoundation/hardhat-toolbox`: It bundles all the commonly used packages and Hardhat plugins.
+- `solidity-coverage`: It is a solidity code coverage plugin for Hardhat.
+- `@openzeppelin/contracts`: It is a secure smart contract library for solidity
+
+# Governance Token Smart Contract
+
+Using the ERC20 token standard, we will develop a governance token for our DAO. Open `Token.sol` and insert the following code:
+
+```js
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+ 
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+ 
+contract Token is ERC20 {
+   
+    constructor() ERC20("Token", "TOKEN") {
+        _mint(msg.sender, 1000000);
+    }
+ 
+}
+```
+
+Please go here to read more about how the above snippet works. We can now go to the smart contract for the DAO.
+DAO Smart Contract
+
+To get things going, we'll construct an empty smart contract called `DAO` in the `DAO.sol` file and import `IERC20.sol` from the openzeppelin library.
+
+```js
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+ 
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+ 
+contract DAO {
+ 
+}
+```
+## Constructor Definition
+
+Now that we have a blank smart contract, we can declare a variable inside of it with the name `token` of type `IERC20`.
+
+The next step is creating the smart contract constructor, which takes the token's address as an argument. We keep the token address as a variable because we want our DAO contract to work with any ERC20 token. We will initialize the Token variable inside the constructor by assigning a value using the `IERC20` interface and handing it an argument of the contract address.
+
+```js
+contract DAO {
+ 
+    IERC20 public token;
+ 
+    constructor(address _token) {
+        token = IERC20(_token);
+    }
+}
+```
+## Defining Proposal
+
+The option to propose and vote on proposals will be available to anyone who has a governance token and chooses to participate in the governance process. We must outline the structure of those proposals. A struct will define the format for each proposal we will create named `Proposal`.
+
+```js
+    struct Proposal {
+        uint256 accept;
+        uint256 reject;
+        uint256 abstain;
+        uint256 deadline;
+        bytes32 title;
+        mapping(address => bool) voted;
+    }
+```
+
+> Note: Everything other than functions goes above the constructor, and all the functions go below the constructor. Just a tip on organizing smart contracts
+Storing Created Proposals
```

Comment:
```suggestion

### Storing Created Proposals
```

Reply:
For keeping formatting consistent, I have made this heading 2.